File: backend/src/OpenDriverHub.Infrastructure/Services/SettingsService.cs
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using OpenDriverHub.Application;
using OpenDriverHub.Domain;

namespace OpenDriverHub.Infrastructure.Services;

/// <summary>
/// Settings de integração. Banco sobrepõe .env/appsettings; valor vazio
/// remove a customização e volta a usar o .env. Segredos são mascarados na leitura.
/// </summary>
public class SettingsService : ISettingsService, ISettingsProvider
{
    private readonly AppDbContext _db;
    private readonly IConfiguration _config;

    public SettingsService(AppDbContext db, IConfiguration config)
    {
        _db = db; _config = config;
    }

    private record Field(string Key, string Label, bool Secret);
    private record Group(string Id, string Name, string Description, string Icon, Field[] Fields);

    private static readonly Group[] Catalog =
    [
        new("whatsapp", "WhatsApp Business",
            "Envia confirmação de compra, código do voucher e lembretes de resgate.",
            "💬",
            [
                new("WhatsApp:Token", "Token da API", true),
                new("WhatsApp:PhoneNumber", "Número (com DDI)", false),
            ]),
        new("mercadopago", "Mercado Pago",
            "Gateway de pagamento (Pix, crédito e débito). Webhook reconcilia o status.",
            "💳",
            [
                new("MercadoPago:AccessToken", "Access Token", true),
                new("MercadoPago:PublicKey", "Public Key", false),
                new("MercadoPago:WebhookSecret", "Webhook Secret", true),
            ]),
        new("email", "E-mail (Gmail)",
            "Conta Gmail para e-mails transacionais (confirmações e relatórios).",
            "✉️",
            [
                new("Email:GmailUser", "Conta Gmail", false),
                new("Email:GmailAppToken", "App Password / Token", true),
                new("Email:FromName", "Nome do remetente", false),
            ]),
        new("groq", "Assistente IA (Groq)",
            "LLM que conduz a conversa do chatbot rumo à conversão. Sem chave, o bot usa o modo local (regras).",
            "🤖",
            [
                new("Groq:ApiKey", "API Key (gsk_...)", true),
                new("Groq:Model", "Modelo", false),
            ]),
    ];

    private static readonly HashSet<string> AllowedKeys =
        Catalog.SelectMany(g => g.Fields.Select(f => f.Key)).ToHashSet();

    public async Task<string?> GetAsync(string key, CancellationToken ct = default)
    {
        var row = await _db.IntegrationSettings
            .FirstOrDefaultAsync(s => s.Key == key, ct);
        if (row is not null && !string.IsNullOrWhiteSpace(row.Value))
            return row.Value;
        var env = _config[key];
        return string.IsNullOrWhiteSpace(env) ? null : env;
    }

    public async Task<List<IntegrationGroupDto>> GetGroupsAsync(CancellationToken ct)
    {
        var rows = await _db.IntegrationSettings.ToListAsync(ct);
        var result = new List<IntegrationGroupDto>();

        foreach (var g in Catalog)
        {
            var fields = new List<IntegrationFieldDto>();
            foreach (var f in g.Fields)
            {
                var dbRow = rows.FirstOrDefault(r => r.Key == f.Key);
                var dbVal = dbRow is not null && !string.IsNullOrWhiteSpace(dbRow.Value)
                    ? dbRow.Value : null;
                var envVal = string.IsNullOrWhiteSpace(_config[f.Key]) ? null : _config[f.Key];
                var effective = dbVal ?? envVal;

                var source = dbVal is not null ? "db" : envVal is not null ? "env" : "unset";
                var preview = effective is null
                    ? ""
                    : f.Secret ? Mask(effective) : effective;

                fields.Add(new IntegrationFieldDto(
                    f.Key, f.Label, f.Secret, effective is not null, preview, source));
            }
            var connected = fields.Where(x =>
                    g.Fields.First(cf => cf.Key == x.Key).Secret)
                .All(x => x.HasValue) && fields.Any(x => x.HasValue);

            result.Add(new IntegrationGroupDto(
                g.Id, g.Name, g.Description, g.Icon, connected, fields));
        }
        return result;
    }

    public async Task UpdateAsync(Guid actorId, UpdateSettingRequest req, CancellationToken ct)
    {
        if (!AllowedKeys.Contains(req.Key))
            throw new AppException("Chave de configuração inválida.", 400);

        var row = await _db.IntegrationSettings
            .FirstOrDefaultAsync(s => s.Key == req.Key, ct);

        if (string.IsNullOrWhiteSpace(req.Value))
        {
            // Vazio → remove customização e volta ao .env.
            if (row is not null) _db.IntegrationSettings.Remove(row);
        }
        else if (row is null)
        {
            _db.IntegrationSettings.Add(new IntegrationSetting
            {
                Key = req.Key,
                Value = req.Value.Trim(),
                UpdatedBy = actorId,
            });
        }
        else
        {
            row.Value = req.Value.Trim();
            row.UpdatedBy = actorId;
            row.UpdatedAt = DateTime.UtcNow;
        }

        _db.AuditLogs.Add(new AuditLog
        {
            ActorId = actorId,
            Action = "settings.update",
            EntityType = "IntegrationSetting",
            EntityId = req.Key,
            // Nunca registramos o valor do segredo, só se foi definido ou limpo.
            PayloadJson = JsonSerializer.Serialize(new
            {
                req.Key,
                cleared = string.IsNullOrWhiteSpace(req.Value),
            }),
        });

        await _db.SaveChangesAsync(ct);
    }

    private static string Mask(string value)
    {
        var v = value.Trim();
        if (v.Length <= 4) return "••••";
        return $"••••••{v[^4..]}";
    }
}

File: backend/src/services/settingsService.ts
import type { IntegrationFieldDto, IntegrationGroupDto, UpdateSettingRequest } from '../dtos/settings.dto.js';
import { AppError } from '../errors.js';
import { prisma } from '../infra/prisma.js';

interface Field {
  key: string;
  label: string;
  secret: boolean;
}
interface Group {
  id: string;
  name: string;
  description: string;
  icon: string;
  fields: Field[];
}

const CATALOG: Group[] = [
  {
    id: 'whatsapp',
    name: 'WhatsApp Business',
    description: 'Envia confirmação de compra, código do voucher e lembretes de resgate.',
    icon: '💬',
    fields: [
      { key: 'WhatsApp:Token', label: 'Token da API', secret: true },
      { key: 'WhatsApp:PhoneNumber', label: 'Número (com DDI)', secret: false },
    ],
  },
  {
    id: 'mercadopago',
    name: 'Mercado Pago',
    description: 'Gateway de pagamento (Pix, crédito e débito). Webhook reconcilia o status.',
    icon: '💳',
    fields: [
      { key: 'MercadoPago:AccessToken', label: 'Access Token', secret: true },
      { key: 'MercadoPago:PublicKey', label: 'Public Key', secret: false },
      { key: 'MercadoPago:WebhookSecret', label: 'Webhook Secret', secret: true },
    ],
  },
  {
    id: 'asaas',
    name: 'Asaas',
    description:
      'Gateway de pagamento com split automático: o líquido cai direto na carteira de cada parceiro (walletId no cadastro). Webhook reconcilia o status.',
    icon: '🏦',
    fields: [
      { key: 'Asaas:ApiKey', label: 'API Key', secret: true },
      { key: 'Asaas:WebhookToken', label: 'Webhook Token', secret: true },
      { key: 'Asaas:Environment', label: 'Ambiente (sandbox/production)', secret: false },
      { key: 'Asaas:DefaultCpfCnpj', label: 'CPF/CNPJ de teste (sandbox)', secret: false },
    ],
  },
  {
    id: 'email',
    name: 'E-mail (Gmail)',
    description: 'Conta Gmail para e-mails transacionais (confirmações e relatórios).',
    icon: '✉️',
    fields: [
      { key: 'Email:GmailUser', label: 'Conta Gmail', secret: false },
      { key: 'Email:GmailAppToken', label: 'App Password / Token', secret: true },
      { key: 'Email:FromName', label: 'Nome do remetente', secret: false },
    ],
  },
  {
    id: 'groq',
    name: 'Assistente IA (Groq)',
    description: 'LLM que conduz a conversa do chatbot rumo à conversão. Sem chave, o bot usa o modo local (regras).',
    icon: '🤖',
    fields: [
      { key: 'Groq:ApiKey', label: 'API Key (gsk_...)', secret: true },
      { key: 'Groq:Model', label: 'Modelo', secret: false },
    ],
  },
];

const ALLOWED_KEYS = new Set(CATALOG.flatMap((g) => g.fields.map((f) => f.key)));

function mask(value: string): string {
  const v = value.trim();
  if (v.length <= 4) return '••••';
  return `••••••${v.slice(-4)}`;
}

function envValue(key: string): string | null {
  const env = process.env[key.replace(/:/g, '__')] ?? process.env[key];
  return env && env.trim() !== '' ? env : null;
}

export async function getGroups(): Promise<IntegrationGroupDto[]> {
  const rows = await prisma.integrationSetting.findMany();
  const result: IntegrationGroupDto[] = [];

  for (const g of CATALOG) {
    const fields: IntegrationFieldDto[] = [];
    for (const f of g.fields) {
      const dbRow = rows.find((r) => r.key === f.key);
      const dbVal = dbRow && dbRow.value.trim() !== '' ? dbRow.value : null;
      const envVal = envValue(f.key);
      const effective = dbVal ?? envVal;
      const source: 'db' | 'env' | 'unset' = dbVal !== null ? 'db' : envVal !== null ? 'env' : 'unset';
      const preview = effective === null ? '' : f.secret ? mask(effective) : effective;
      fields.push({ key: f.key, label: f.label, secret: f.secret, hasValue: effective !== null, preview, source });
    }
    const connected =
      fields.filter((x) => g.fields.find((cf) => cf.key === x.key)!.secret).every((x) => x.hasValue) &&
      fields.some((x) => x.hasValue);
    result.push({ id: g.id, name: g.name, description: g.description, icon: g.icon, connected, fields });
  }
  return result;
}

export async function updateSetting(actorId: string, req: UpdateSettingRequest): Promise<void> {
  if (!ALLOWED_KEYS.has(req.key)) throw new AppError('Chave de configuração inválida.', 400);

  const row = await prisma.integrationSetting.findUnique({ where: { key: req.key } });

  if (!req.value || req.value.trim() === '') {
    // Vazio → remove customização e volta ao .env.
    if (row) await prisma.integrationSetting.delete({ where: { key: req.key } });
  } else if (!row) {
    await prisma.integrationSetting.create({ data: { key: req.key, value: req.value.trim(), updatedBy: actorId } });
  } else {
    await prisma.integrationSetting.update({
      where: { key: req.key },
      data: { value: req.value.trim(), updatedBy: actorId, updatedAt: new Date() },
    });
  }

  await prisma.auditLog.create({
    data: {
      actorId,
      action: 'settings.update',
      entityType: 'IntegrationSetting',
      entityId: req.key,
      // Nunca registramos o valor do segredo, só se foi definido ou limpo.
      payloadJson: JSON.stringify({ key: req.key, cleared: !req.value || req.value.trim() === '' }),
    },
  });
}
